File: silencendo/TMplatform/src/ingestion/chunkManager.ts
import * as fs from 'fs';
import * as path from 'path';
import { Chunk } from './types';

const CHUNKS_FILE = path.join('.bot', 'chunks.json');

export interface ChunkStorage {
  chunks: Chunk[];
}

export class ChunkManager {
  private storage: ChunkStorage;

  constructor() {
    this.storage = { chunks: [] };
    this.loadChunks();
  }

  private ensureBotDir() {
    if (!fs.existsSync('.bot')) {
      fs.mkdirSync('.bot', { recursive: true });
    }
  }

  private loadChunks(): void {
    this.ensureBotDir();
    
    if (fs.existsSync(CHUNKS_FILE)) {
      try {
        const data = JSON.parse(fs.readFileSync(CHUNKS_FILE, 'utf8'));
        this.storage = data;
      } catch (error) {
        console.error('Error loading chunks:', error);
        // Use default state if loading fails
      }
    } else {
      // Initialize with empty state
      this.saveChunks();
    }
  }

  public saveChunks(): void {
    this.ensureBotDir();
    fs.writeFileSync(CHUNKS_FILE, JSON.stringify(this.storage, null, 2));
  }

  public addChunks(chunks: Chunk[]): void {
    this.storage.chunks.push(...chunks);
    this.saveChunks();
  }

  public getChunksBySource(sourceId: string): Chunk[] {
    return this.storage.chunks.filter(chunk => chunk.sourceId === sourceId);
  }

  public getAllChunks(): Chunk[] {
    return this.storage.chunks;
  }

  public removeChunksBySource(sourceId: string): void {
    this.storage.chunks = this.storage.chunks.filter(chunk => chunk.sourceId !== sourceId);
    this.saveChunks();
  }
  
  public clearChunks(): void {
    this.storage.chunks = [];
    this.saveChunks();
  }

  public getChunksByIds(ids: string[]): Chunk[] {
    return this.storage.chunks.filter(chunk => ids.includes(chunk.chunkId));
  }
}

File: silencendo/TMplatform/ingestion/chunkManager.go
package ingestion

import (
	"encoding/json"
	"io/ioutil"
	"os"
)

const CHUNKS_FILE = ".bot/chunks.json"

type ChunkStorage struct {
	Chunks []Chunk `json:"chunks"`
}

type ChunkManager struct {
	storage ChunkStorage
}

func NewChunkManager() *ChunkManager {
	cm := &ChunkManager{
		storage: ChunkStorage{
			Chunks: []Chunk{},
		},
	}
	cm.loadChunks()
	return cm
}

func (cm *ChunkManager) ensureBotDir() {
	if _, err := os.Stat(".bot"); os.IsNotExist(err) {
		os.MkdirAll(".bot", 0755)
	}
}

func (cm *ChunkManager) loadChunks() {
	cm.ensureBotDir()

	if _, err := os.Stat(CHUNKS_FILE); err == nil {
		data, err := ioutil.ReadFile(CHUNKS_FILE)
		if err == nil {
			err = json.Unmarshal(data, &cm.storage)
			if err != nil {
				// Use default state if loading fails
				cm.saveChunks()
			}
		} else {
			// Use default state if loading fails
			cm.saveChunks()
		}
	} else {
		// Initialize with empty state
		cm.saveChunks()
	}
}

func (cm *ChunkManager) saveChunks() {
	cm.ensureBotDir()
	data, err := json.MarshalIndent(cm.storage, "", "  ")
	if err == nil {
		ioutil.WriteFile(CHUNKS_FILE, data, 0644)
	}
}

func (cm *ChunkManager) AddChunks(chunks []Chunk) {
	cm.storage.Chunks = append(cm.storage.Chunks, chunks...)
	cm.saveChunks()
}

func (cm *ChunkManager) GetChunksBySource(sourceId string) []Chunk {
	var result []Chunk
	for _, chunk := range cm.storage.Chunks {
		if chunk.SourceID == sourceId {
			result = append(result, chunk)
		}
	}
	return result
}

func (cm *ChunkManager) GetAllChunks() []Chunk {
	return cm.storage.Chunks
}

func (cm *ChunkManager) RemoveChunksBySource(sourceId string) {
	var newChunks []Chunk
	for _, chunk := range cm.storage.Chunks {
		if chunk.SourceID != sourceId {
			newChunks = append(newChunks, chunk)
		}
	}
	cm.storage.Chunks = newChunks
	cm.saveChunks()
}

func (cm *ChunkManager) ClearChunks() {
	cm.storage.Chunks = []Chunk{}
	cm.saveChunks()
}

func (cm *ChunkManager) GetChunksByIds(ids []string) []Chunk {
	var result []Chunk
	idMap := make(map[string]bool)
	for _, id := range ids {
		idMap[id] = true
	}

	for _, chunk := range cm.storage.Chunks {
		if idMap[chunk.ChunkID] {
			result = append(result, chunk)
		}
	}
	return result
}
